Improve the UX of the SMTP Section

Ridhwana opened the pull request:
This branch is not names aptly as it metamorphosed into a different feature.

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

In this PR:

I rename the SMTP Header on the config page and make some copy changes. I also make a significant UX change that determines whether to display a 'use your own server' toggle, some description text and the SMTP Form based on whether the user has a Sendgrid API Key and/or custom SMTP Settings. 

This is how the feature should behave: 

| Sendgrid API Key | Custom SMTP Settings | Show Toggle Element  | Text description                                                                                                                                                                                                                                  | Action                                                                               |
|------------------|----------------------|----------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|
| no               | no                   | no                   | none                                                                                                                                                                                                                                              | Show the SMTP Form                                                                   |
| yes              | no                   | yes                  | As a Forem Cloud client, we provide an email server managed by the Forem team. All settings are managed by us and the from and reply email addresses are set as noreply@forem.com. . However, you can override this to use your own email server. | When the user toggles the box, the SMTP Form will show up                            |
| no               | yes                  | no                   | none                                                                                                                                                                                                                                              | Show the SMTP Form                                                                   |
| yes              | yes                  | yes                  | none                                                                                                                                                                                                                                              | A user can un toggle and it should clear out the form and revert to the sengrid key. |


You can read more about the discussion [here](https://github.com/forem/forem/issues/16180).

In a subsequent PR I'll add a reply_to address, a from_email address and a contact email address. 

You can find a demo video here:
https://www.loom.com/share/472c958d890140dfa13400ee3f054a10

## Related Tickets & Documents
https://github.com/forem/forem/issues/16180

## QA Instructions, Screenshots, Recordings

You can find a demo video here:
https://www.loom.com/share/472c958d890140dfa13400ee3f054a10

In the table above where it says `SENDGRID_API_KEY` needs to be set. You will need to set your  `SENDGRID_API_KEY` in env.yml. When you add or remove your ENV key you will need to restart the server.

For the  'yes'.  values in `Custom SMTP Settings` column, you will need to set the minimum fields required for SMTP: address, username, password.  

Navigate to /admin/customization/config` and ensure that what you see in the Email Server Section is based on the table in the description.
 
### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

citizen428 commented on app/javascript/admin/controllers/config_controller.js:
```diff
@@ -16,6 +16,8 @@ const emailAuthModalBody = `
   <p>If you disable Email address as a registration option, people cannot create an account with their email address.</p>
   <p>However, people who have already created an account using their email address can continue to login.</p>`;
 
+// Note: In an effort to move away from Stimulus and create consistency across the codebase
```
<3 

Also, totally tangential suggestion. It's customary to capitalize NOTE, TODO, FIXME, etc. because many/most coding editors will specifically highlight those and/or have extensions for treating them like todo lists, etc.

citizen428 commented on app/javascript/packs/admin/config/smtp.js:
```diff
@@ -0,0 +1,27 @@
+    ];
+
+    for (let i = 0; i < inputs.length; i++) {
+      if (!haveDefaultValues.some((el) => inputs.item(i).name.includes(el))) {
```
I will forever wonder why they couldn't also add `Array.prototype.none` together with `every` and `some`.

Ridhwana replied:
Yip will do - this was still a Draft, was just getting it working and then planning to polish up before review.

citizen428 commented on app/views/admin/settings/forms/_smtp.html.erb:
```diff
@@ -1,22 +1,46 @@
+<%= javascript_packs_with_chunks_tag "admin/config/smtp", defer: true %>
+<% only_sendgrid_enabled = ForemInstance.sendgrid_enabled? && !Settings::SMTP.provided_minimum_settings? %>
```
Could it make sense to move this whole check into `ForemInstance.only_sendgrid_enabled?` I'm just not super comfortable with views knowing **anything** about business logic.

Ridhwana replied:
Yip will do - this was still a Draft, was just getting it working and then planning to polish up before review.

citizen428 replied:
Yes, sorry, I was a bit too early here 😊

Ridhwana commented on app/controllers/stories/feeds_controller.rb:
```diff
@@ -65,8 +65,8 @@ def signed_in_base_feed
     def signed_out_base_feed
-      strategy = AbExperiment.get(experiment: AbExperiment::CURRENT_FEED_STRATEGY_EXPERIMENT, controller: self, user: current_user,
-                                  default_value: AbExperiment::ORIGINAL_VARIANT)
+      strategy = AbExperiment.get(experiment: AbExperiment::CURRENT_FEED_STRATEGY_EXPERIMENT, controller: self,
```
Some Rubocop fixes for LayoutLength

Ridhwana commented on app/models/settings/smtp.rb:
```diff
@@ -2,11 +2,12 @@ module Settings
     self.table_name = :settings_smtp
 
-    OPTIONS = %i[address port authentication user_name password domain].freeze
+    OPTIONS = %i[user_name password address authentication domain port].freeze
```
changed the order of the fields to be a bit more friendly

Ridhwana commented on app/views/admin/settings/forms/_smtp.html.erb:
```diff
@@ -1,22 +1,47 @@
   <div class="card mt-3" id="smtp-section">
     <%= render partial: "admin/shared/card_header",
-               locals: { header: "SMTP Settings", state: "collapse", target: "smtpSettingsBodyContainer", expanded: false } %>
+               locals: { header: "Email Server Settings (SMTP)", state: "collapse", target: "smtpSettingsBodyContainer", expanded: false } %>
```
renamed the SMTP section to a more understandable term

juliannatetreault commented on spec/system/admin/config/admin_updates_smtp_settings_spec.rb:
```diff
@@ -0,0 +1,111 @@
+      end
+    end
+
+    it "shows the SMTP Form" do
```
Quick, non-blocking suggestion: can we please aggregate the failures here and anywhere else where we have multiple expectations within a single `it` block in this spec file (lines `81` and `104`)?
```suggestion
    it "shows the SMTP Form", :aggregate_failures do
```

juliannatetreault commented on spec/system/admin/config/admin_updates_smtp_settings_spec.rb:
```diff
@@ -0,0 +1,111 @@
+      end
+    end
+
+    it "shows the SMTP Form" do
```
Quick, non-blocking suggestion: same as above comment!
```suggestion
    it "shows the SMTP Form", :aggregate_failures do
```

juliannatetreault reviewed (commented):
I left a few, non-blocking comments, but otherwise, this is looking great so far! ✨

juliannatetreault reviewed (approved):
Just two more nit-picky comments (that I couldn't point to directly in the PR files, so it may be better to address them in a follow-up PR), otherwise this feature works as expected and looks good to go! 😊 

**1.** Quick nitpick: Can we please update the "User name" header to read "Username" instead, as shown in the screenshot below?
<img width="941" alt="Screen Shot 2022-02-07 at 8 10 56 AM" src="https://user-images.githubusercontent.com/32834804/152815586-fcb84a19-a648-40c8-94ce-1ec47462fabe.png">

**2.** Final, quick nitpick: Additionally, can we please update the "ie." on line `7`, as shown [here](https://github.com/forem/forem/blob/08f6440dea7be5bfef8c44369242450019c17d74/app/lib/constants/settings/smtp.rb#L7), to read "i.e." rather than "ie."?

nickytonline commented on app/javascript/packs/admin/config/smtp.js:
```diff
@@ -0,0 +1,25 @@
+const checkbox = document.getElementById('settings_smtp_own_email_server');
+
+if (checkbox) {
+  checkbox.addEventListener('change', function () {
+    const customSMTPSection = document.getElementsByClassName(
+      'js-custom-smtp-section',
+    )[0];
+
+    if (this.checked) {
+      customSMTPSection.classList.remove('hidden');
+    } else {
+      // when the user indicates that they do not want to use their own server
+      // we clear the form values except for those fields that have a default value.
+      const inputs = customSMTPSection.getElementsByTagName('input');
+      const haveDefaultValues = ['authentication'];
+
+      for (let i = 0; i < inputs.length; i++) {
+        if (!haveDefaultValues.some((el) => inputs.item(i).name.includes(el))) {
+          inputs.item(i).value = '';
+        }
+      }
+      customSMTPSection.classList.add('hidden');
+    }
+  });
+}
```
Refactor (blocking): Use [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) instead as well as `event.target` instead of `this`.

```suggestion
document
  .getElementById('settings_smtp_own_email_server')
  ?.addEventListener('change', (event) => {
    const customSMTPSection = document.getElementsByClassName(
      'js-custom-smtp-section',
    )[0];

    if (event.target.checked) {
      customSMTPSection.classList.remove('hidden');
    } else {
      // when the user indicates that they do not want to use their own server
      // we clear the form values except for those fields that have a default value.
      const inputs = customSMTPSection.getElementsByTagName('input');
      const haveDefaultValues = ['authentication'];

      for (let i = 0; i < inputs.length; i++) {
        if (!haveDefaultValues.some((el) => inputs.item(i).name.includes(el))) {
          inputs.item(i).value = '';
        }
      }
      customSMTPSection.classList.add('hidden');
    }
  });

```

Refactor (nitpick): Prefer restructuring

```diff
-   const customSMTPSection = document.getElementsByClassName(
+   const [ customSMTPSection ] = document.getElementsByClassName(
      'js-custom-smtp-section',
-   )[0];
+   );
```

Refactor (nitpick): Prefer `for`/`of`

```diff
- for (let i = 0; i < inputs.length; i++) {
+ for (const input of inputs) {
- if (!haveDefaultValues.some((el) => inputs.item(i).name.includes(el))) {
+ if (!haveDefaultValues.some((el) => input.name.includes(el))) {
-    inputs.item(i).value = '';
+    input.value = '';
  }
}
```

Ridhwana replied:
Thanks Nick,I'll update these!

Ridhwana commented:
> Just two more nit-picky comments (that I couldn't point to directly in the PR files, so it may be better to address them in a follow-up PR), otherwise this feature works as expected and looks good to go! 😊
> 
> **1.** Quick nitpick: Can we please update the "User name" header to read "Username" instead, as shown in the screenshot below? <img alt="Screen Shot 2022-02-07 at 8 10 56 AM" width="941" src="https://user-images.githubusercontent.com/32834804/152815586-fcb84a19-a648-40c8-94ce-1ec47462fabe.png">

@juliannatetreault this is generated automatically based on the name of the setting "user_name". We also setting these values in the DOM through a loop so it would be a lot more effort that it may potentially be worth to try and fix this.

> 
> **2.** Final, quick nitpick: Additionally, can we please update the "ie." on line `7`, as shown [here](https://github.com/forem/forem/blob/08f6440dea7be5bfef8c44369242450019c17d74/app/lib/constants/settings/smtp.rb#L7), to read "i.e." rather than "ie."?

Will do! Thank you!

aitchiss commented on cypress/integration/seededFlows/adminFlows/config/emailServerSettingsSection.spec.js:
```diff
@@ -0,0 +1,57 @@
+        cy.visit('/admin/customization/config');
+        cy.findByTestId('emailServerSettings').as('emailServerSettings');
+
+        cy.get('@emailServerSettings')
```
Since most of the action takes place within this element, you could do

```js
cy.get('@emailServerSettings').within(() => {
  // Everything in here is scoped to inside that element
  cy.findByText('Email Server Settings (SMTP)').click();
  cy.get('#settings_smtp_user_name').clear();
})
```

Ridhwana replied:
✅

aitchiss commented on cypress/integration/seededFlows/adminFlows/config/emailServerSettingsSection.spec.js:
```diff
@@ -0,0 +1,57 @@
+          .click();
+
+        cy.get('@emailServerSettings')
+          .get('#settings_smtp_user_name')
```
can we target by the label text - e.g. `cy.findByLabelText('label for this input')`? Same question for the other input fields below - it makes our tests a bit more robust / closer to how a user experiences the page 🙂

Ridhwana replied:
Thanks for picking this up @aitchiss - I knew this, not sure why I used ID's.  It also helped bring to light that the for labels weren't matching, which I fixed.

Ridhwana commented on cypress/integration/seededFlows/adminFlows/config/emailServerSettingsSection.spec.js:
```diff
@@ -0,0 +1,57 @@
+    cy.fixture('users/adminUser.json').as('user');
+
+    cy.get('@user').then((user) => {
+      cy.loginUser(user);
```
✅

Ridhwana commented:
Re-requested reviews as I've addressed feedback, but also moving this into draft so that it doesn't get accidentally merged whilst I complete part 2 and work on a post explaining the changes + documentation
